pkg/maps/metricsmap/metricsmap.go

```diff
@@ -212,69 +206,48 @@ func updateMetric(getCounter func() (prometheus.Counter, error), newValue float6
 // updatePrometheusMetrics checks the metricsmap key value pair
 // and determines which prometheus metrics along with respective labels
 // need to be updated.
-func updatePrometheusMetrics(key *Key, val *Value) {
+func updatePrometheusMetrics(key *Key, values *Values) {
+	// Metrics is a per-CPU map so we first need to aggregate the
+	// different entries that make up a value.
+	var packets, bytes uint64
+	for _, value := range *values {
+		packets += value.Count
+		bytes += value.Bytes
+	}
+
 	updateMetric(func() (prometheus.Counter, error) {
 		if key.IsDrop() {
 			return metrics.DropCount.GetMetricWithLabelValues(key.DropForwardReason(), key.Direction())
 		}
 		return metrics.ForwardCount.GetMetricWithLabelValues(key.Direction())
-	}, val.CountFloat())
+	}, float64(packets))
 
 	updateMetric(func() (prometheus.Counter, error) {
 		if key.IsDrop() {
 			return metrics.DropBytes.GetMetricWithLabelValues(key.DropForwardReason(), key.Direction())
 		}
 		return metrics.ForwardBytes.GetMetricWithLabelValues(key.Direction())
-	}, val.bytesFloat())
+	}, float64(bytes))
 }
 
 // SyncMetricsMap is called periodically to sync off the metrics map by
 // aggregating it into drops (by drop reason and direction) and
 // forwards (by direction) with the prometheus server.
 func SyncMetricsMap(ctx context.Context) error {
-	entry := make([]Value, possibleCpus)
-	file := bpf.MapPath(MapName)
-
-	var err error
-	metricsMap := bpf.GetMap(file)
-	if metricsMap == nil {
-		// Open the map and leave it open, since SyncMetricsMap is called
-		// periodically and it makes sense to use an already opened map rather
-		// than opening the map again and again.
-		// This also prevents the constant registration and unregistration of the
-		// Map.
-		metricsMap, err = bpf.OpenMap(file)
-
-		if err != nil {
-			return fmt.Errorf("Unable to open metrics map: %s", err)
-		}
+	callback := func(key bpf.MapKey, values bpf.MapValue) {
+		// Increment Prometheus metrics here.
+		updatePrometheusMetrics(key.(*Key), values.(*Values))
 	}
 
-	var key, nextKey Key
-	for {
-		err := bpf.GetNextKey(metricsMap.GetFd(), unsafe.Pointer(&key), unsafe.Pointer(&nextKey))
-		if err != nil {
-			break
-		}
-		err = bpf.LookupElement(metricsMap.GetFd(), unsafe.Pointer(&nextKey), unsafe.Pointer(&entry[0]))
-		if err != nil {
-			return fmt.Errorf("unable to lookup metrics map: %s", err)
-		}
-
-		// cannot use `range entry` since, if the first value for a particular
-		// CPU is zero, it never iterates over the next non-zero value.
-		for i := 0; i < possibleCpus; i++ {
-			// Increment Prometheus metrics here.
-			updatePrometheusMetrics(&nextKey, &entry[i])
```

Comment:
Since you added a comment to that regard, I thought that the bug was that we simply forgot to iterate on per-CPU values, but it seems we were already doing that here :point_up: 
Am I missing something?

Reply:
This is slightly subtle as I'm changing the logic from:

```go
func updatePrometheusMetrics(value) {
  updateMetric(value.Packets)
  updateMetric(value.Bytes)
}

func SyncMetrics() {
  for _, value := range values {
    updatePrometheusMetrics(value)
  }
}
```

to:

```go
func updatePrometheusMetrics(values) {
  for _, value := range values {
    packets += value.Packets
    bytes   += value.Bytes
  }

  updateMetrics(packets)
  updateMetrics(bytes)
}


func SyncMetrics() {
  updatePrometheusMetrics(values)
}
```

What `updateMetrics` does is:
https://github.com/cilium/cilium/blob/8bf3ed83cfbbbf8c9da2d7c57c1b8debecc6b35d/pkg/maps/metricsmap/metricsmap.go#L199-L210

so:

* diff the `packets`/`bytes` value it receive with the one stored in Prometheus
* increment the Prometheus counter by the delta it just calculated

Since the logic in `updateMetrics` is not linear, it will work correctly only if the `packets`/`bytes` counter it receives is the _sum_ of all the per-CPU values for a given key (rather than what we are doing now, which is calling `updateMetrics()` on every single per-CPU value for a given key)

Reply:
Ah, got it. Thanks for the explanation!

It would probably make sense to explain that in the commit message, or to break it down in two commits to separate the fix from the refactoring.

Reply:
Good shout, let me split the commit